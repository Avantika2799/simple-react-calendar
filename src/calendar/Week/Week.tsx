import React, { FC, SyntheticEvent } from 'react'

import { IDateRange, IWeekDayRenderProps, IWeekRenderProps } from '../../@types'
import WeekDay from '../WeekDay'

export type Props = {
  blockClassName: string
  customRender?: IWeekRenderProps
  getDayEachDay: ({ start, end }: IDateRange) => Date[]
  getDayFormatted: (date: Date) => string
  getDayIsCurrentMonth: (date: Date) => boolean
  getDayIsDisabled: (date: Date) => boolean
  getDayIsHighlighted: (date: Date) => boolean
  getDayIsNextMonth: (date: Date) => boolean
  getDayISOFormatter: (date: Date) => string
  getDayIsSelectable: (date: Date) => boolean
  getDayIsSelected: (date: Date) => boolean
  getDayIsSelectedEnd: (date: Date) => boolean
  getDayIsSelectedStart: (date: Date) => boolean
  getDayIsToday: (date: Date) => boolean
  getDayIsWeekend: (date: Date) => boolean
  handleOnDayClick: (event: SyntheticEvent<HTMLButtonElement>) => void
  handleOnDayMouseEnter: (event: SyntheticEvent<HTMLButtonElement>) => void
  handleOnDisabledDayClick: (event: SyntheticEvent<HTMLButtonElement>) => void
  handleOnDisabledDayMouseEnter: (event: SyntheticEvent<HTMLButtonElement>) => void
  renderWeekDay?: IWeekDayRenderProps
  weekEnd: Date
  weekNumber: number
  weekStart: Date
}

const Week: FC<Props> = (props) => {
  const {
    blockClassName,
    customRender,
    getDayEachDay,
    getDayFormatted,
    getDayIsCurrentMonth,
    getDayIsDisabled,
    getDayIsHighlighted,
    getDayIsNextMonth,
    getDayISOFormatter,
    getDayIsSelectable,
    getDayIsSelected,
    getDayIsSelectedEnd,
    getDayIsSelectedStart,
    getDayIsToday,
    getDayIsWeekend,
    handleOnDayClick,
    handleOnDayMouseEnter,
    handleOnDisabledDayClick,
    handleOnDisabledDayMouseEnter,
    renderWeekDay,
    weekEnd,
    weekNumber,
    weekStart,
  } = props

  const children = getDayEachDay({ start: weekStart, end: weekEnd }).map((day: Date) => {
    const date = getDayISOFormatter(day)
    const isCurrentMonth = getDayIsCurrentMonth(day)
    const isDisabled = getDayIsDisabled(day)
    const isNextMonth = !isCurrentMonth && getDayIsNextMonth(day)
    const isSelectable = getDayIsSelectable(day)
    const isWorkDay = !getDayIsWeekend(day)

    return (
      <WeekDay
        blockClassName={blockClassName}
        customRender={renderWeekDay}
        date={day}
        getDayFormatted={getDayFormatted}
        handleOnClick={isSelectable ? handleOnDayClick : handleOnDisabledDayClick}
        handleOnEnter={isSelectable ? handleOnDayMouseEnter : handleOnDisabledDayMouseEnter}
        isCurrentMonth={getDayIsCurrentMonth(day)}
        isDisabled={isDisabled}
        isHighlighted={getDayIsHighlighted(day)}
        isMonthNext={isNextMonth}
        isMonthPrev={!isCurrentMonth && !isNextMonth}
        isNonSelectable={!isSelectable}
        ISODate={date}
        isSelectable={isSelectable}
        isSelected={getDayIsSelected(day)}
        isSelectionEnd={getDayIsSelectedEnd(day)}
        isSelectionStart={getDayIsSelectedStart(day)}
        isToday={getDayIsToday(day)}
        isWeekend={!isWorkDay}
        isWorkday={isWorkDay}
        key={date}
      />
    )
  })

  if (customRender) {
    return customRender(props)
  }

  return (
    <div className={`${blockClassName}-week`}>
      <div className={`${blockClassName}-week_number`}>{weekNumber}</div>
      <div className={`${blockClassName}-week_days`}>{children}</div>
    </div>
  )
}

Week.displayName = 'Week'

export default Week
