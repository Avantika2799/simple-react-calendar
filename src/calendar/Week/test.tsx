import { cleanup, fireEvent, render, RenderResult } from '@testing-library/react'
import React from 'react'

import { IDateRange, ISelectionRange } from '../../@types'
import {
  getDayEachDay,
  getDayFormatted,
  getDayIsCurrentMonth,
  getDayIsDisabled,
  getDayIsHighlighted,
  getDayIsNextMonth,
  getDayISOFormatter,
  getDayIsSame,
  getDayIsSelectable,
  getDayIsSelected,
  getDayIsSelectedEnd,
  getDayIsSelectedStart,
  getDayIsWeekend,
} from '../helper'
import Week, { Props } from './Week'

afterEach(cleanup)

describe('Week', () => {
  let api: RenderResult,
    activeMonth: Date,
    disabledIntervals: IDateRange[],
    maxDate: Date,
    minDate: Date,
    mockedCustomRender: any,
    mode: 'range' | 'single',
    props: Props,
    selected: ISelectionRange,
    today: Date

  beforeEach(() => {
    activeMonth = new Date(2019, 10, 1)
    disabledIntervals = []
    maxDate = new Date(2019, 10, 26)
    minDate = new Date(2019, 10, 20)
    mode = 'range'
    selected = { inProgress: false, start: new Date(2019, 10, 28), end: new Date(2019, 10, 29) }
    today = new Date(2019, 10, 27)

    props = {
      blockClassName: 'test-class-name',
      getDayEachDay,
      getDayFormatted,
      getDayIsCurrentMonth: (date) => getDayIsCurrentMonth({ activeMonth, date }),
      getDayIsDisabled: (date) => getDayIsDisabled({ date, disabledIntervals }),
      getDayIsHighlighted: (date) => getDayIsHighlighted({ date, highlightedIntervals: [] }),
      getDayIsNextMonth: (date) => getDayIsNextMonth({ activeMonth, date }),
      getDayISOFormatter,
      getDayIsSelectable: (date) => getDayIsSelectable({ date, minDate, maxDate, disabledIntervals }),
      getDayIsSelected: (date) => getDayIsSelected({ date, mode, selected }),
      getDayIsSelectedEnd: (date) => getDayIsSelectedStart({ date, mode, selected }),
      getDayIsSelectedStart: (date) => getDayIsSelectedEnd({ date, mode, selected }),
      getDayIsToday: (date) => getDayIsSame({ date, dateToCompare: today }),
      getDayIsWeekend: (date) => getDayIsWeekend({ date }),
      handleOnDayClick: jest.fn(),
      handleOnDayMouseEnter: jest.fn(),
      handleOnDisabledDayClick: jest.fn(),
      handleOnDisabledDayMouseEnter: jest.fn(),
      weekEnd: new Date(2019, 11, 1),
      weekNumber: 48,
      weekStart: new Date(2019, 10, 25),
    }
  })

  describe('action handlers', () => {
    beforeEach(() => {
      api = render(<Week {...props} />)
    })

    test('invoke `#handleOnDayClick`', () => {
      const { getByText } = api

      fireEvent.click(getByText(/26/i))

      expect(props.handleOnDayClick).toHaveBeenCalledTimes(1)
      expect(props.handleOnDisabledDayClick).not.toHaveBeenCalled()
    })

    test('invoke `#handleOnDisabledDayClick`', () => {
      const { getByText } = api

      fireEvent.click(getByText(/27/i))

      expect(props.handleOnDisabledDayClick).toHaveBeenCalledTimes(1)
      expect(props.handleOnDayClick).not.toHaveBeenCalled()
    })

    test('invoke `#handleOnDayMouseEnter`', () => {
      const { getByText } = api

      fireEvent.mouseEnter(getByText(/26/i))

      expect(props.handleOnDayMouseEnter).toHaveBeenCalledTimes(1)
      expect(props.handleOnDisabledDayMouseEnter).not.toHaveBeenCalled()
    })

    test('invoke `#handleOnDisabledDayMouseEnter`', () => {
      const { getByText } = api

      fireEvent.mouseEnter(getByText(/27/i))

      expect(props.handleOnDisabledDayMouseEnter).toHaveBeenCalledTimes(1)
      expect(props.handleOnDayMouseEnter).not.toHaveBeenCalled()
    })
  })

  describe('when render prop has not been passed', () => {
    beforeEach(() => {
      api = render(<Week {...props} />)
    })

    test('default render', () => {
      const { container } = api

      expect(container).toMatchSnapshot()
    })
  })

  describe('when render prop has been passed', () => {
    beforeEach(() => {
      mockedCustomRender = jest.fn(() => null)

      props = {
        ...props,
        customRender: mockedCustomRender,
      }
      api = render(<Week {...props} />)
    })

    test('render prop has been called properly', () => {
      expect(props.customRender).toHaveBeenCalledTimes(1)
      expect(props.customRender).toHaveBeenCalledWith(props)
    })
  })
})
