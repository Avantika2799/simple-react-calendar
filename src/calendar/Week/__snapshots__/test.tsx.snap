// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Week when render prop has not been passed default render 1`] = `
<div>
  <div
    class="test-class-name-week"
  >
    <div
      class="test-class-name-week_number"
    >
      48
    </div>
    <div
      class="test-class-name-week_days"
    >
      <div
        class="test-class-name-day_wrapper"
      >
        <button
          class="test-class-name-day is-current_month is-selectable is-working_day"
          type="button"
          value="2019-11-25"
        >
          25
        </button>
      </div>
      <div
        class="test-class-name-day_wrapper"
      >
        <button
          class="test-class-name-day is-current_month is-selectable is-working_day"
          type="button"
          value="2019-11-26"
        >
          26
        </button>
      </div>
      <div
        class="test-class-name-day_wrapper"
      >
        <button
          class="test-class-name-day is-current_month is-not_selectable is-today is-working_day"
          type="button"
          value="2019-11-27"
        >
          27
        </button>
      </div>
      <div
        class="test-class-name-day_wrapper"
      >
        <button
          class="test-class-name-day is-current_month is-end_selection is-not_selectable is-selected is-working_day"
          type="button"
          value="2019-11-28"
        >
          28
        </button>
      </div>
      <div
        class="test-class-name-day_wrapper"
      >
        <button
          class="test-class-name-day is-current_month is-not_selectable is-selected is-start_selection is-working_day"
          type="button"
          value="2019-11-29"
        >
          29
        </button>
      </div>
      <div
        class="test-class-name-day_wrapper"
      >
        <button
          class="test-class-name-day is-current_month is-not_selectable is-weekend"
          type="button"
          value="2019-11-30"
        >
          30
        </button>
      </div>
      <div
        class="test-class-name-day_wrapper"
      >
        <button
          class="test-class-name-day is-next_month is-not_selectable is-weekend"
          type="button"
          value="2019-12-01"
        >
          1
        </button>
      </div>
    </div>
  </div>
</div>
`;
